src/render.c

```diff
@@ -59,15 +333,128 @@ void animate(int fps, anim_function renderer,
+        // escape sequence, it will write the response to the standard input of
+        // the program, terminating with a BEL.
+        sprintf(buffer, "\033]4;%d;?\007", i + 1);
+        err = query_terminal(buffer, buffer, ESCAPE_CODE_SIZE);
+        if(err != 0){
+            fprintf(stderr, "Error reading from buffer: %d\n", err);
+            err = ERR_QUERY_UNSUPPORTED;
+            goto error;
+        }
+
+        // xterm and urxvt give different responses to the query. Xterm is
+        // sensible, and includes the colour index in the response
+        // (\033]4;i;rgb:), but urxvt doesn't (\033]4;rgb:).
+        // We get around this by explicitly inserting the index here if it
+        // is not present.
+        int ignore;
+        if(sscanf(buffer, "\033]%d;rgb", &ignore)) {
+            char *start = strstr(buffer, "rgb");
+
+            int escape_sz = snprintf(NULL, 0, "\033]4;%d;%s", i + 1, start);
+            if(escape_sz < 0) {
+                fprintf(stderr, "Error calculating length of escape buffer\n");
+                err = ERR_OUT_OF_MEMORY;
+                goto error;
+            }
+
+            char *escape_buf = calloc(escape_sz + 1, 1);
+            if(escape_sz < 0) {
+                perror("Error allocating escape code memory.");
+                err = ERR_OUT_OF_MEMORY;
+                goto error;
+            }
+
+            sprintf(escape_buf, "\033]4;%d;%s", i + 1, start);
+            escape_codes[i] = escape_buf;
+        }else{
+            escape_codes[i] = strdup(buffer);
+            if(!escape_codes[i]) {
+                perror("Error allocating escape code memory.");
+                err = ERR_OUT_OF_MEMORY;
+                goto error;
+            }
+        }
+    }
+    return 0;
+
+error:
+    for(i = i - 1; i > 0; i--)
+        free(escape_codes[i]);
+    free(escape_codes);
+    return err;
+}
+
+
+/** Print backup escape codes to terminal. */
+void restore_colors(struct color_backup *backup) {
+    for(int i=0; i < backup->num_colors; i++) {
+        putp(backup->escape_codes[i]);
+    }
+}
+
+/** Free the list of escape codes used for restoring colours. */
+void free_colors(struct color_backup *backup) {
+    for(int i=0; i < backup->num_colors; i++) {
+        free(backup->escape_codes[i]);
+    }
+}
+
+/**
+ * Write a query sequence to the terminal, and wait for a response.
+ * The reponse is written to `buffer`.
+ */
+int query_terminal(const char *escape, char *buffer, int bufsz) {
+    putp(escape);
+    int c;
+    int i = 0;
+    timeout(100);
+    do {
+        if(i >= bufsz)
+            return -1;
+
+        c = getch();
+        if(c == ERR)
+            return -2;
+
+        buffer[i++] = c;
+    } while(c != '\7');
+    buffer[i] = '\0';
+    return 0;
+}
+
 void render_pipe(struct pipe *p, char **trans, char **pipe_chars,
         int old_state, int new_state){
 
     move(p->y, p->x);
-    attron(COLOR_PAIR(p->colour));
+    attr_set(A_NORMAL, 1, &p->colour);
```

Comment:
```c
In file included from src/render.c:6:0:
src/render.c: In function 'render_pipe':
/usr/include/ncursesw/curses.h:1301:13: warning: right-hand operand of comma expression has no effect [-Wunused-value]
 #define wattr_set(win,a,p,opts)  (((win) \
                                  ~~~~~~~~~
        ? ((win)->_attrs = ((a) & ~A_COLOR), \
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           (win)->_color = (p)) \
           ~~~~~~~~~~~~~~~~~~~~~~
        : OK), \
        ~~~~~^~~
       OK)
       ~~~
/usr/include/ncursesw/curses.h:1206:26: note: in expansion of macro 'wattr_set'
 #define attr_set(a,c,o)  wattr_set(stdscr,(a),(c),(o))
                          ^~~~~~~~~
src/render.c:453:5: note: in expansion of macro 'attr_set'
     attr_set(A_NORMAL, 1, &p->colour);
     ^~~~~~~~
```

I think all warnings should be treated as errors, that is `-Werror`, but `configure.ac` has something for a specific warning and I don't know how we can work around these three diagnoses (`-Wunused-value` and two notes), when  the problem is in ncurses' code.  Well, by we, I meant you will figure it out, I am sure. :)

Reply:
I wasn't able to replicate this on my system, using either gcc 7.3.1 or clang 6.0.0 with ncurses 6.1. I don't know about clang, but I'm fairly sure gcc will suppress warnings from system library headers. Could you give me some info about your system so I can try and reproduce this?

Regarding errors as warnings, `configure.ac` was supposed to set `-Werror` for non-release builds---that is, builds that aren't exactly on a tag. However, I forgot to add `--tags` to the `git describe` command line that gets the version, and the GitHub uses lightweight tags for releases, so `git describe --always` was just giving the abbreviated commit ID! This obviously doesn't contain a hyphen, so `configure` thought it was always building a release. Frankly, I have no idea how I manged to build the 1.0.0 tarball with `make dist`: if I check out `v1.0.0` now and run `make dist` I get `pipes-c-d161b9c.tar.gz`. I will fix this in a separate PR immediately and rebase this PR onto it.

Finally, I agree that *most* warnings should be treated as errors, but not all. I currently have a couple of unused parameters that I don't care about in functions that are called by function pointers, but those functions need to have a consistent signature. I'll leave the compiler flags as they are for the moment: it should be more useful after fixing `configure.ac`.

Reply:
Your mention about ncurses was very helpful, I was using ncurses 6.0 and so I updated to 6.1, now GCC 6.4.0 compiles it without any warnings ([`curses.h`](https://github.com/mirror/ncurses/commit/3eda6f30a84d53844d2ebceadb457e2e7e9cfbf3#diff-76c2c1047ac829d06e97e1ca5d8af0b2R1340) **and** the five [`-Wundef`](https://github.com/mirror/ncurses/commit/90d42867e1296bc79021006a92032c76e59068b6#diff-02df643b0c44408698c57a2812a6867bR44) warnings)

This is made clear to me, I think we should *require* ncurses 6.1+ even though pipes.c can be compiled and run with ncurses 6.0 and those warnings.

Reply:
Forgot to also mention that clang 3.9.1 doesn't show the warnings on curses.h of ncurses 6.0.

(3.9.1 is still in Gentoo Portage tree as stable, the latest stable is 5.0.1, but I don't have it installed.  Travis CI has 5, but ncurses version is unknown from the log)